howso/client/api.py

```diff
@@ -0,0 +1,138 @@
+from __future__ import annotations
+
+from collections.abc import Mapping
+from functools import lru_cache
+import json
+from pathlib import Path
+import typing as t
+from uuid import uuid4
+
+from typing_extensions import NotRequired, TypeAlias, TypedDict
+
+from amalgam.api import Amalgam
+from .exceptions import HowsoError
+
+DEFAULT_ENGINE_PATH = Path(__file__).parent.parent.joinpath("howso-engine")
+
+SchemaTypeOption: TypeAlias = t.Literal["any", "assoc", "boolean", "list", "number", "string", "null"]
+"""Enum of valid Engine types."""
+
+SchemaType: TypeAlias = t.Union[SchemaTypeOption, list[SchemaTypeOption]]
+"""Valid schema type options."""
+
+TypeDefinition: TypeAlias = t.Union[SchemaType, "Ref", "Schema", "AnyOf"]
+"""Any valid type definition. Type(s), Reference, Schema, or Schema composition."""
+
+
+class Ref(TypedDict):
+    """Reference to another schema."""
+
+    ref: str
+    description: NotRequired[str | None]
+    required: NotRequired[bool]
+    default: NotRequired[t.Any]
+
+
+class AnyOf(TypedDict):
+    """An "OR" schema composition."""
+
+    any_of: TypeDefinition
+    description: NotRequired[str | None]
+    required: NotRequired[bool]
+    default: NotRequired[t.Any]
+
+
+class Schema(TypedDict):
+    """A definition of a schema in Engine."""
+
+    type: SchemaType
+    description: NotRequired[str]
+    required: NotRequired[bool]
+    default: NotRequired[t.Any]
+    enum: NotRequired[list[int | float | str]]
+    min: NotRequired[int | float]
+    max: NotRequired[int | float]
+    exclusive_min: NotRequired[int | float]
+    exclusive_max: NotRequired[int | float]
+    min_size: NotRequired[int]
+    max_size: NotRequired[int]
+    values: NotRequired[TypeDefinition]
+    min_indices: NotRequired[int]
+    max_indices: NotRequired[int]
+    indices: NotRequired[Mapping[str, TypeDefinition]]
+    additional_indices: NotRequired[TypeDefinition | bool]
+    dynamic_indices: NotRequired[TypeDefinition]
+
+
+class LabelDefinition(TypedDict):
+    """A definition to an Engine label."""
+
+    parameters: Mapping[str, TypeDefinition] | None
+    returns: NotRequired[TypeDefinition | None]
+    description: NotRequired[str | None]
+    attribute: NotRequired[bool]
+    use_active_session: NotRequired[bool]
+    long_running: NotRequired[bool]
+    idempotent: NotRequired[bool]
+    statistically_idempotent: NotRequired[bool]
+    read_only: NotRequired[bool]
+    payload: NotRequired[bool]
+
+
+class EngineApi(TypedDict):
+    """The Howso Engine Api documentation object."""
+
+    labels: Mapping[str, LabelDefinition]
+    """Map of Engine label name to label definition."""
+
+    schemas: Mapping[str, Schema | Ref | AnyOf]
+    """Mapping of schema name to schema definition."""
+
+    description: str
+    """Description of the API."""
+
+
+@lru_cache(16)
```

Comment:
That's an oddly specific size for the cache. Any particular reason it's 16 and not 10, or 20?

Reply:
Just explicitly set to a lower value so the cache doesn't grow too large. I don't generally expect this to really ever exceed 1 value in the cache.